src/virtio/sound.c

```diff
@@ -0,0 +1,851 @@
+#include "sound.h"
+#include "config.h"
+#include "microkit.h"
+#include "virq.h"
+#include "virtio/mmio.h"
+#include "virtio/virtq.h"
+#include <sddf/sound/queue.h>
+
+#define DEBUG_SOUND
+
+#if defined(DEBUG_SOUND)
+#define LOG_SOUND(...) do{ printf("VIRTIO(SOUND): "); printf(__VA_ARGS__); }while(0)
+#else
+#define LOG_SOUND(...) do{}while(0)
+#endif
+
+#define LOG_SOUND_ERR(...) do{ printf("VIRTIO(SOUND)|ERROR: "); printf(__VA_ARGS__); }while(0)
+
+#define CONTROLQ 0
+#define EVENTQ 1
+#define TXQ 2
+#define RXQ 3
+
+#define MIN(a, b) ((a) < (b) ? (a) : (b))
+
+static inline struct virtio_snd_device *device_state(struct virtio_device *dev)
+{
+    return (struct virtio_snd_device *)dev->device_data;
+}
+
+static void virtio_snd_mmio_reset(struct virtio_device *dev)
+{
+    LOG_SOUND("Resetting virtIO sound device\n");
+
+    for (int i = 0; i < VIRTIO_SND_NUM_VIRTQ; i++) {
+        dev->vqs[i].ready = 0;
+        dev->vqs[i].last_idx = 0;
+    }
+}
+
+static int virtio_snd_mmio_get_device_features(struct virtio_device *dev, uint32_t *features)
+{
+    switch (dev->data.DeviceFeaturesSel) {
+        case 0:
+            // virtIO sound does not define any features
+            *features = 0;
+            break;
+        case 1:
+            *features = BIT_HIGH(VIRTIO_F_VERSION_1);
+            break;
+        default:
+            LOG_SOUND_ERR("driver sets DeviceFeaturesSel to 0x%x, which doesn't make sense\n", dev->data.DeviceFeaturesSel);
+            return 0;
+    }
+    return 1;
+}
+
+static int virtio_snd_mmio_set_driver_features(struct virtio_device *dev, uint32_t features)
+{
+    int success;
+    switch (dev->data.DriverFeaturesSel) {
+        // feature bits 0 to 31
+        case 0:
+            success = (features == 0);
+            break;
+        // features bits 32 to 63
+        case 1:
+            success = (features == BIT_HIGH(VIRTIO_F_VERSION_1));
+            break;
+        default:
+            LOG_SOUND_ERR("driver sets DriverFeaturesSel to 0x%x, which doesn't make sense\n", dev->data.DriverFeaturesSel);
+            success = 0;
+    }
+
+    if (success) {
+        dev->data.features_happy = 1;
+    }
+
+    return success;
+}
+
+static int virtio_snd_mmio_get_device_config(struct virtio_device *dev, uint32_t offset, uint32_t *ret_val)
+{
+    // @alexbr: duplicated from block. can we abstract this?
+    struct virtio_snd_device *state = device_state(dev);
+    uintptr_t config_base_addr = (uintptr_t)&state->config;
+    uintptr_t config_field_offset = (uintptr_t)(offset - REG_VIRTIO_MMIO_CONFIG);
+    uint32_t *config_field_addr = (uint32_t *)(config_base_addr + config_field_offset);
+    *ret_val = *config_field_addr;
+
+    return 1;
+}
+
+static int virtio_snd_mmio_set_device_config(struct virtio_device *dev, uint32_t offset, uint32_t val)
+{
+    LOG_SOUND_ERR("Not allowed to change sound config.");
+    return 0;
+}
+
+// static const char *queue_name[] = { "controlq", "eventq", "txq", "rxq", NULL };
```

Comment:
Is this still necessary?

Reply:
Removed in https://github.com/au-ts/libvmm/commit/547db9f51de00395bb355f56855758b1bd4c12bb